common/archiver/s3store/historyArchiver.go

```diff
@@ -0,0 +1,333 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// S3 History Archiver will archive workflow histories to amazon s3
+
+package s3store
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/service/config"
+)
+
+const (
+	// URIScheme is the scheme for the s3 implementation
+	URIScheme               = "s3"
+	errBucketNotExists      = "requested bucket does not exist"
+	errEncodeHistory        = "failed to encode history batches"
+	errWriteKey             = "failed to write history to s3"
+	defaultBlobstoreTimeout = 10 * time.Second
+	targetHistoryBlobSize   = 2 * 1024 * 1024 // 2MB
+)
+
+var (
+	errNoBucketSpecified = errors.New("no bucket specified")
+	errEmptyAwsRegion    = errors.New("empty aws region")
+)
+
+type (
+	historyArchiver struct {
+		container *archiver.HistoryBootstrapContainer
+		s3cli     s3iface.S3API
+		// only set in test code
+		historyIterator archiver.HistoryIterator
+		config          *config.S3Archiver
+	}
+
+	getHistoryToken struct {
+		CloseFailoverVersion int64
+		NextBatchIdx         int
+	}
+)
+
+// NewHistoryArchiver creates a new archiver.HistoryArchiver based on s3
+func NewHistoryArchiver(
+	container *archiver.HistoryBootstrapContainer,
+	config *config.S3Archiver,
+) (archiver.HistoryArchiver, error) {
+	return newHistoryArchiver(container, config, nil)
+}
+
+func newHistoryArchiver(
+	container *archiver.HistoryBootstrapContainer,
+	config *config.S3Archiver,
+	historyIterator archiver.HistoryIterator,
+) (*historyArchiver, error) {
+	if len(config.Region) == 0 {
+		return nil, errEmptyAwsRegion
+	}
+	s3Config := &aws.Config{
+		Endpoint:         config.Endpoint,
+		Region:           aws.String(config.Region),
+		S3ForcePathStyle: aws.Bool(config.S3ForcePathStyle),
+		MaxRetries:       aws.Int(0),
+	}
+	sess, err := session.NewSession(s3Config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &historyArchiver{
+		container:       container,
+		s3cli:           s3.New(sess),
+		historyIterator: historyIterator,
+	}, nil
+}
+func (h *historyArchiver) Archive(
```

Comment:
Please emit metrics in this method. 

MetricScope: `HistoryArchiverScope`
MetricNames:
```
        CadenceLatency
	HistoryArchiverArchiveNonRetryableErrorCount
	HistoryArchiverArchiveTransientErrorCount
	HistoryArchiverArchiveSuccessCount
	HistoryArchiverTotalUploadSize
	HistoryArchiverHistorySize
```

In your case upload size and history size will be the same since it's always a full upload.

Reply:
Done!